crates/bevy_render/src/pipeline/compute_pipeline_compiler.rs

```diff
@@ -0,0 +1,173 @@
+use super::{ComputePipelineDescriptor, DynamicBinding, ShaderSpecialization, SpecializedShader};
+use crate::{
+    renderer::RenderResourceContext,
+    shader::{Shader, ShaderSource},
+};
+use bevy_asset::{Assets, Handle};
+use bevy_property::Properties;
+use once_cell::sync::Lazy;
+use std::collections::HashMap;
+
+#[derive(Clone, Eq, PartialEq, Debug, Properties)]
+pub struct ComputePipelineSpecialization {
+    pub shader_specialization: ShaderSpecialization,
+    pub dynamic_bindings: Vec<DynamicBinding>,
+}
+
+impl Default for ComputePipelineSpecialization {
+    fn default() -> Self {
+        Self {
+            shader_specialization: Default::default(),
+            dynamic_bindings: Default::default(),
+        }
+    }
+}
+
+impl ComputePipelineSpecialization {
+    pub fn empty() -> &'static ComputePipelineSpecialization {
+        pub static EMPTY: Lazy<ComputePipelineSpecialization> =
+            Lazy::new(ComputePipelineSpecialization::default);
+        &EMPTY
+    }
+}
+
+struct ComputeSpecializedPipeline {
+    pipeline: Handle<ComputePipelineDescriptor>,
+    specialization: ComputePipelineSpecialization,
+}
+
+#[derive(Default)]
+pub struct ComputePipelineCompiler {
+    specialized_shaders: HashMap<Handle<Shader>, Vec<SpecializedShader>>,
+    specialized_pipelines:
+        HashMap<Handle<ComputePipelineDescriptor>, Vec<ComputeSpecializedPipeline>>,
+}
+
+impl ComputePipelineCompiler {
+    // TODO: Share some of this with PipelineCompiler.
```

Comment:
Theres more code duplication here than I would like. Theres a lot of "implementation divergence risk" as it stands. id rather do one of these:

1. add an enum or something that wraps the two descriptors and use that during compilation.
2. have one descriptor for the "common stuff", but then have an enum inside that for compute stuff vs rasterization stuff.
3. make PipelineCompiler generic on descriptors and implement a Trait that extracts the required data.

Reply:
I think the enum approach makes sense. Compute pipelines are very very small and only require bind group layouts.

Reply:
Works for me!